### PR TITLE
feat: use LoggingLogBucketRef

### PR DESCRIPTION
**Pull request by acpana**
Start using the LoggingLogBucketRef!

### CODE REVIEW

File: pkg/controller/direct/logging/logmetric_controller.go [hunk 1]
```diff
@@ -150,6 +150,24 @@ func (a *logMetricAdapter) Create(ctx context.Context, u *unstructured.Unstructu
...
 		return fmt.Errorf("filter is empty")
 	}
+	if a.desired.Spec.LoggingLogBucketRef != nil {
+		bucketName := a.desired.Spec.LoggingLogBucketRef.External
+		if bucketName != "" {
+			// todo acpana support resolving non external ref for loggingLogBucketRef
+
+			// validate the the bucket ref external is well formatted
+			// eg: projects/my-project/locations/global/buckets/my-bucket
+			parts := strings.Split(bucketName, "/")
```
Comment (yuwenma) [inline pkg/controller/direct/logging/logmetric_controller.go:160]: I think we will want some general util functions to validate and parse the full resource name (basically the "parent + resource" form), like
- `projects/{PROJECT}/locations/{LOCATION_ID}/{RESOURCE_NAMES}/{RESOURCE_ID}`
- `projects/{PROJECT}/{RESOURCE_NAMES}/{RESOURCE_ID}`
and probably some others. 

How about a new file `pkg/controller/direct/parse/resourceid.go` for that purpose?  Open to other dir/file naming ideas.

File: pkg/controller/direct/logging/logmetric_controller.go [hunk 2]
```diff
@@ -150,6 +150,24 @@ func (a *logMetricAdapter) Create(ctx context.Context, u *unstructured.Unstructu
 	if filter == "" {
 		return fmt.Errorf("filter is empty")
 	}
+	if a.desired.Spec.LoggingLogBucketRef != nil {
+		bucketName := a.desired.Spec.LoggingLogBucketRef.External
+		if bucketName != "" {
+			// todo acpana support resolving non external ref for loggingLogBucketRef
```
Comment (yuwenma) [inline pkg/controller/direct/logging/logmetric_controller.go:156]: super nit: suggest using the todo format, like `TODO(apana)`.

File: pkg/controller/direct/logging/logmetric_controller.go [hunk 3]
```diff
@@ -150,6 +150,24 @@ func (a *logMetricAdapter) Create(ctx context.Context, u *unstructured.Unstructu
...
+
+			// validate the the bucket ref external is well formatted
+			// eg: projects/my-project/locations/global/buckets/my-bucket
+			parts := strings.Split(bucketName, "/")
+			if len(parts) != 6 || parts[0] != "projects" || parts[2] != "locations" || parts[4] != "buckets" {
+				return fmt.Errorf("bucketName %q is not in the format projects/PROJECT_ID/locations/LOCATION_ID/buckets/BUCKET_ID", bucketName)
+			}
+
+			// validate that the bucket is in the same project
+			if a.parentID != fmt.Sprintf("projects/%s", parts[1]) { // todo acpana rebase on main and rework
```
Comment (yuwenma) [inline pkg/controller/direct/logging/logmetric_controller.go:166]: Nice!

File: pkg/controller/direct/logging/logmetric_controller.go (hunk 1 above)
Comment (acpana) [reply pkg/controller/direct/logging/logmetric_controller.go:160]: There has been some chatter about this in the past. I think we can move towards having some `parser.go`.  I can follow on this PR with more usage (outside of this controller). cc @justinsb

File: pkg/controller/direct/logging/logmetric_controller.go (hunk 2 above)
Comment (acpana) [reply pkg/controller/direct/logging/logmetric_controller.go:156]: I can start using this moving forward. It would be nice to have a linter for this hehe.

**Review by justinsb — APPROVED** [review]

**Comment by justinsb** [comment]
LGTM!

/approve
/lgtm

File: pkg/test/resourcefixture/testdata/basic/logging/v1beta1/logginglogmetric/logbucketmetric/create.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    sku: "EXTRACT(jsonPayload.id)"
+  bucketOptions:
+    linearBuckets:
+      numFiniteBuckets: 3
+      width: 3.5
+      offset: 1.5
+  projectRef:
+    external: "projects/${projectId}"
+  loggingLogBucketRef:
+    external: "projects/${projectId}/locations/global/buckets/logginglogbucket-dep-${uniqueId}"
```
Comment (jingyih) [inline pkg/test/resourcefixture/testdata/basic/logging/v1beta1/logginglogmetric/logbucketmetric/create.yaml:51]: Do we support the following?
```
LoggingLogBucketRef:
  name:
```

Comment (acpana) [reply pkg/test/resourcefixture/testdata/basic/logging/v1beta1/logginglogmetric/logbucketmetric/create.yaml:51]: I think anything that is not `external` will come as a follow on for now.

Comment (jingyih) [reply pkg/test/resourcefixture/testdata/basic/logging/v1beta1/logginglogmetric/logbucketmetric/create.yaml:51]: Oh I see. This is a new field we are adding. That makes sense. Thank you!

**Pull request merged**